File: nombre/strobogrammatique/strobogrammatique.py
# -----------------------------------------------
# strobogrammatique
# -----------------------------------------------
# Un nombre strobogrammatique peut se lire de la
# même façon après avoir subi une rotation de 180
# degrés.
# -----------------------------------------------


# -----------------------------------------------
# Version courte
# -----------------------------------------------

def strobogrammatique(nombre):
    chifs = ["0", "1", "8", "6", "9"]
    corrs = ["0", "1", "8", "9", "6"]
    
    if isinstance(nombre, int):
        nombre = str(nombre)
    
    for i in range(len(nombre) // 2 + 1):
        chif = nombre[i]
        corr = nombre[len(nombre) - 1 - i]
        
        if chif not in chifs or \
            corrs[chifs.index(chif)] != corr:
            return False
            
    return True



# -----------------------------------------------
# Application
# -----------------------------------------------

nombres = [88, 133, 96, 101]
for nb in nombres:
    res = strobogrammatique(nb)
    strobo = "strobogrammatique";
    if res:
        print(f"{nb} est {strobo}.")
    else:
        print(f"{nb} n'est pas {strobo}.")



# -----------------------------------------------
# Version golf
# -----------------------------------------------

S=lambda n:all(str(n)[i]in"01869"and"01896"["018\
69".index(str(n)[i])]==str(n)[~i]for i in range(\
len(str(n))//2+1))



# -----------------------------------------------
# Version detaillee
# -----------------------------------------------

def strobogrammatique_detaillee(nombre):
    strobo = {
        "0" : "0",
        "1" : "1",
        "8" : "8",
        "6" : "9",
        "9" : "6"
        }
    
    if type(nombre) is int:
        nombre = str(nombre)
        
    gauche = 0
    droite = len(nombre) - 1
    
    while droite - gauche >= 0:
        premier = nombre[gauche]
        dernier = nombre[droite]
        if (
            not premier in strobo.keys() or
            not dernier in strobo.keys() or
            premier != strobo[dernier]
            ):
            return False
        
        droite -= 1
        gauche += 1
        
    return True



# -----------------------------------------------
# Version commentee
# -----------------------------------------------

def strobogrammatique_commentee(nombre):
    
    # Définir un dictionnaire de correspondance.
    strobo = {
        "0" : "0",
        "1" : "1",
        "8" : "8",
        "6" : "9",
        "9" : "6"
        }
    
    # Vérifier si le type du nombre est entier.
    if type(nombre) is int:
        
        # Convertir le nombre en chaîne.
        nombre = str(nombre)
    
    # Initialiser les index.
    gauche = 0
    droite = len(nombre) - 1
    
    # Tant que la différence est supérieure ou
    # égale à 0.
    while droite - gauche >= 0:
        
        # Obtenir le premier chiffre.
        premier = nombre[gauche]
        
        # Obtenir le dernier chiffre.
        dernier = nombre[droite]
        
        # Si,
        if (
            # le premier n'est pas dans le
            # dictionnaire.
            not premier in strobo.keys() or
            
            # le dernier n'est pas dans le
            # dictionnaire.
            not dernier in strobo.keys() or
            
            # si le premier et le dernier sont
            # différents.
            premier != strobo[dernier]
            ):
            
            # Retourner faux.
            return False
        
        # Décrémenter les index.
        droite -= 1
        gauche += 1
    
    # Retourner le resultat.
    return True









#

File: nombre/strobogrammatique/strobogrammatique.js
//-----------------------------------------------
// strobogrammatique 
/*----------------------------------------------- 
/ Un nombre strobogrammatique peut se lire de la
/ même façon après avoir subi une rotation de 180
/ degrés.
/----------------------------------------------*/


//-----------------------------------------------
// Version courte
//-----------------------------------------------

function strobogrammatique(nombre) {
    const strobo = {
        "0": "0", "1": "1", "8": "8", "6": "9",
        "9": "6" };

    nombre = nombre.toString();

    let deb = 0;
    let fin = nombre.length - 1;

    for (deb, fin; fin >= deb; deb++, fin--) {

        let premier = nombre[deb];
        let dernier = nombre[fin];

        if (!(premier in strobo &&
            strobo[premier] === dernier)) {
                return false;
            }
    }

    return true;
}



//-----------------------------------------------
// Application
//-----------------------------------------------

let nombres = [88, 133, 96, 101];
for (let nb of nombres) {
    let res = strobogrammatique(nb);
    let strobo = "strobogrammatique"
    if (res) {
        console.log(`${nb} est ${strobo}`);
    }
    else {
        console.log(`${nb} n'est pas ${strobo}`);
    }
}



//-----------------------------------------------
// Version golf
//-----------------------------------------------

const S=n=>{const s={"0":"0","1":"1","8":"8","6":
"9","9":"6"};return(n=n.toString(),([...n].every(
(c,i)=>c in s&&s[c]===n[n.length-1-i])))};



//-----------------------------------------------
// Version detaillee
//-----------------------------------------------

function strobogrammatique_detaillee(nombre) {

    const strobo = {
        "0": "0",
        "1": "1",
        "8": "8",
        "6": "9",
        "9": "6"
    };

    if (typeof nombre === 'number') {
        nombre = nombre.toString();
    }

    let gauche = 0;
    let droite = nombre.length - 1;

    while (droite - gauche >= 0) {

        let premier = nombre[gauche];
        let dernier = nombre[droite];

        if (
            !(premier in strobo) ||
            !(dernier in strobo) ||
            premier !== strobo[dernier]
        ) {
            return false;
        }

        droite -= 1;
        gauche += 1;
    }

    return true;
}

//-----------------------------------------------
// Version commentee
//-----------------------------------------------

function strobogrammatique_commentee(nombre) {

    // Définir un dictionnaire de correspondance.
    const strobo = {
        "0": "0",
        "1": "1",
        "8": "8",
        "6": "9",
        "9": "6"
    };

    // Vérifier si nombre est un entier.
    if (typeof nombre === 'number') {

        // Convertir le nombre en string.
        nombre = nombre.toString();
    }

    // Initialiser les index.
    let gauche = 0;
    let droite = nombre.length - 1;

    // Tant que la différence est positive,
    while (droite - gauche >= 0) {

        // Obtenir le premier chiffre.
        let premier = nombre[gauche];

        // Obtenir le dernier chiffre.
        let dernier = nombre[droite];

        // Si,
        if (
            // le premier n'est pas dans le
            // dictionnaire,
            !(premier in strobo) ||

            // le dernier n'est pas dans le
            // dictionnaire,
            !(dernier in strobo) ||

            // si le premier et le dernier sont
            // différents.
            premier !== strobo[dernier]
        ) {

            // Alors, retourner faux.
            return false;
        }

        // Décrémenter les index.
        droite -= 1;
        gauche += 1;
    }

    // Retourner le resultat.
    return true;
}









//